### DISCUSSION TITLE
Best practice for coding instrument functionality that you can’t test

### DISCUSSION THREAD

**Discussion by mcdo0486** [pymeasure/pymeasure#717] [Ideas]
In coding up more pymeasure instrument drivers I have run into cases where I can code functionality into the driver that I can’t test.

As a hypothetical example, there is an instrument that measures wind speed and lightning strikes.
The base model of the instrument comes without the lightning module, which is a $50K upgrade, and I have the base model.

I code up the `wind_speed = Instrument.control()` property and test it out. It works great.

Now with the instrument manual it would be simple enough to code up the lightning strike instrument property, however I can’t test it and this property requires passing in 1000 amps to the lightning rod, or whatever. A mistake in the implementation can cause serious problems for the instrument or user.

Should I include code for instrument functionality I can’t test?

I think there are four options:

1. Keep status quo
2. Don’t code any instrument functionality you can’t test yourself
3. Include “Tested” or “Not Tested” in the docstrings for all coded functionality
4. Include “Property is UNTESTED” at the end of the docstring for coded functionality you couldn’t test

I think option 4 is the most practical. Properties can be assumed to be tested unless it is noted that “Property is UNTESTED” in the docstring.

**Comment by BenediktBurger** [comment]
We have tests, which do not require the device connected (see [protocol tests](https://pymeasure.readthedocs.io/en/latest/dev/adding_instruments.html#protocol-tests)). You can use a manual to write those tests and verify your code with those tests. Additionally, you can mark the code with a comment as yet untested, such that someone, who has that device may test it. If you fear, that it may harm, you might raise a warning or (if really necessary) even an exception.

**Reply by bilderbuchi** [reply]
Of course, the first of those methods (protocol tests) is preferred, because we can then continually make sure that the situation is unchanged. 😀

**Reply by mcdo0486** [reply]
That is a good point and I should have differentiated between physically testing and protocol testing. Protocol testing is great for catching errors and breaking changes to the code scaffolding of pymeasure. 

In the hypothetical I could code up `inst.lightening_strikes` and write up a protocol test that would work for that code, however I haven't physically tested it. Maybe there are details in the manual I missed, like setup commands or something, that wouldn't be obvious until the functionality was physically tested.

I think physically testing is different from protocol testing, and both are needed. Or rather, if you can't physically test some functionality, you should put "Property is UNTESTED" in the docstring as a best practice. I think that will be more obvious than a source code comment.

The hypothetical is extreme case that emphasizes safety risk, and the use of a warning or exception could be added to the best practice. I think the more likely benefit for establishing a best practice about being clear about what functionality hasn't been physically tested will be the bug prevention for people looking to use that functionality for the first time.

**Reply by BenediktBurger** [reply]
I see your point in marking untested methods.
On the other hand, it may clutter the code, albeit not too much.
Now I tend to recommend marking code that way, but I'm not yet hooked.

**Reply by bilderbuchi** [reply]
I'm not sold on this, as it will produce a lot of clutter I fear.
I think on the case of hazardous functionality, a pragmatic approach would be to not implement it if you cannot physically test it. This is safe, avoids the need to call out a warning, and the next person who needs it will naturally be in a position to test it. 
I estimate that those properties will probably be few. Most will be for some setting etc, and them not working should just cause a bug report.

**Reply by LongnoseRob** [reply]
> I'm not sold on this, as it will produce a lot of clutter I fear. I think on the case of hazardous functionality, a pragmatic approach would be to not implement it if you cannot physically test it. This is safe, avoids the need to call out a warning, and the next person who needs it will naturally be in a position to test it. I estimate that those properties will probably be few. Most will be for some setting etc, and them not working should just cause a bug report.

One could implement the features and comment-out the block with a TODO comment that because of missing HW this part of the implementation could not be verified..

```
# TODO: Tests with real HW needed 
# the following block has not been verified due to missing HW(-option)
# use at your own risk
# If you are able to test the code with the correct hardware, please do so and if successfull file a PR at ...

# lightening_curr = Instrument.measurement(...)
```

**Reply by mcdo0486** [reply]
I appreciate the discussion on the topic, and I think it speaks to the need to develop some best practices in the documentation. 

I agree that hazardous functionality probably shouldn't be implemented if it can't be tested, as a best practice. As Rob said, it can be coded but commented out and not available in the instument API documentation. I think that is a best practice that should be included in the documentation. Maybe this can be added to a collection of "Best Practices" that can go in ["Adding instruments" ](https://pymeasure.readthedocs.io/en/latest/dev/adding_instruments.html)

For non-hazardous functionality that can't be tested, I don't think adding the three words "Property is UNTESTED" to the end of a docstring will clutter up the codebase. As mentioned the times when a feature can't be tested won't be too often, but setting up a best practice can help people that are developing new devices and avoid bugs by users looking to onboard onto pymeasure.

**Reply by untzag** [reply]
Just a random thought, an `@untested` decorator could be made. Using such a decorator, users could flip a switch to turn all the untested methods into exceptions.

**Reply by bilderbuchi** [reply]
A decorator sounds like an elegant solution, I had another one in mind recently, too (@deprecated).

> Maybe this can be added to a collection of "Best Practices" that can go in ["Adding instruments" ](https://pymeasure.readthedocs.io/en/latest/dev/adding_instruments.html)

I mean, most of that document is a collection of "best practices", so it's not a large leap to add some points about untested features. :D

**Reply by BenediktBurger** [reply]
> I mean, most of that document is a collection of "best practices", so it's not a large leap to add some points about untested features. :D

I think the "Tests" section might be the best place: Adding another subsection, that you should test the code with a device or mark untested parts (the criteria have to be determined).

> A decorator sounds like an elegant solution.

I agree. That way it is easy to remove as well.

> (@deprecated)

Good idea. That decorator could contain to parameters: Version and the warning message. It could also add the deprecation note (with the warning message as content) to the docstring. How to include it in Sphinx, I do not know.

**Reply by mcdo0486** [reply]
`@untested` would be a convenient way to flag that functionality, but you can't use decorators before assigning a property via the static_method `Instrument.control()`:
```
# not possible
@untested
wind_speed = Instrument.control()
```
Let me know if I'm wrong about that. Maybe `untested` can be passed as an argument to `Instrument.control()` instead?

**Reply by bilderbuchi** [reply]
Decorators can be used in a function form, too, so `deprecated(Instrunent.control())` should work.

**Reply by mcdo0486** [reply]
> Decorators can be used in a function form, too, so `deprecated(Instrunent.control())` should work.

I'd call that a function not a decorator, but I tie the `@deprecated` syntax to decorators. But I see that functionally  `deprecated(Instrument.control())` does the same thing a decorator wrapper.

**Reply by bilderbuchi** [reply]
> > Decorators can be used in a function form, too, so `deprecated(Instrunent.control())` should work.
> 
> I'd call that a function not a decorator, but I tie the `@deprecated` syntax to decorators. But I see that functionally `deprecated(Instrument.control())` does the same thing a decorator wrapper.

Well, that's what a decorator *is*, see e.g. https://docs.python.org/3/glossary.html#term-decorator

**Reply by mcdo0486** [reply]
> > > Decorators can be used in a function form, too, so `deprecated(Instrunent.control())` should work.
> > 
> > 
> > I'd call that a function not a decorator, but I tie the `@deprecated` syntax to decorators. But I see that functionally `deprecated(Instrument.control())` does the same thing a decorator wrapper.
> 
> Well, that's what a decorator _is_, see e.g. https://docs.python.org/3/glossary.html#term-decorator

Yea, but who doesn't like syntactic sugar? I can't get enough of it.

**Reply by bilderbuchi** [reply]
That's not what I said? You said "I'd call that a function not a decorator", and that's what I wanted to provide more context for.

**Comment by LongnoseRob** [comment]
> In coding up more pymeasure instrument drivers I have run into cases where I can code functionality into the driver that I can’t test.
> 
> As a hypothetical example, there is an instrument that measures wind speed and lightning strikes. The base model of the instrument comes without the lightning module, which is a $50K upgrade, and I have the base model.
...
>     4. Include “Property is UNTESTED” at the end of the docstring for coded functionality you couldn’t test
> 
> 
> I think option 4 is the most practical. Properties can be assumed to be tested unless it is noted that “Property is UNTESTED” in the docstring.

Idea 4 could be ehanced with the `icon` [sphinx extesion](https://github.com/sphinx-contrib/icon):
we could then add the following to the docstring of untested entities:
```

"""...
:icon:`fal fa-siren-on`
 This property has not been tested, use on own risk! 
 Feedback wanted on successfull tests.
"""

```
This would show [this icon](https://fontawesome.com/v5/icons/siren-on?f=classic&s=light) and a text block

**Reply by bilderbuchi** [reply]
No need for an extension, we can just append a [warning](https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-warning).

**Reply by msmttchr** [reply]
> > In coding up more pymeasure instrument drivers I have run into cases where I can code functionality into the driver that I can’t test.
> > As a hypothetical example, there is an instrument that measures wind speed and lightning strikes. The base model of the instrument comes without the lightning module, which is a $50K upgrade, and I have the base model.
> > ...
> > 4. Include “Property is UNTESTED” at the end of the docstring for coded functionality you couldn’t test
> > I think option 4 is the most practical. Properties can be assumed to be tested unless it is noted that “Property is UNTESTED” in the docstring.
> 
> Idea 4 could be ehanced with the `icon` [sphinx extesion](https://github.com/sphinx-contrib/icon): we could then add the following to the docstring of untested entities:
> 
> ```
> 
> """...
> :icon:`fal fa-siren-on`
>  This property has not been tested, use on own risk! 
>  Feedback wanted on successfull tests.
> """
> ```
> 
> This would show [this icon](https://fontawesome.com/v5/icons/siren-on?f=classic&s=light) and a text block

I think, it's clear what we want to achieve with marker "untested" in whatever form it will land.
What's not clear to me is what we do imply for the code is not marked "untested" ? I am under the impression we want to give some sort of guarantee. To me, none of the code provided by pymeasure can be considered safe or guaranteed in anyway and users use it at their own risk.

**Reply by bilderbuchi** [reply]
That's an excellent point! 
The only thing I could imagine is having test coverage of all non-"untested" code! This sounds unrealistic to me as it would ~double effort, and test coverage alone guarantees ~nothing.

**Reply by BenediktBurger** [reply]
We cannot (and should not) give guarantees, but normally we are confident, that the code works.
In some cases, you code along the manual, but are unsure, whether it works as expected (due to an obscure mechanism, combination...).
In that case, the `untested` flag could be understood as a request for tests / feedback.

**Reply by mcdo0486** [reply]
> I think, it's clear what we want to achieve with marker "untested" in whatever form it will land. What's not clear to me is what we do imply for the code is not marked "untested" ? I am under the impression we want to give some sort of guarantee. To me, none of the code provided by pymeasure can be considered safe or guaranteed in anyway and users use it at their own risk.

As I mentioned in another thread, the liability for coding or executing procedures is always on the person that is running the code. Pymeasure doesn't give any guarantees, that's part of the MIT license. But we may be able decrease the number of headaches with untested functionality.

Initially I suggested  “Property is UNTESTED” at the end of an untested function's docstring. Other suggestions include wrapping with a simple decorator to programmatically raise errors and modify docstring in a standard way. 

Pros:
1. Make it clear what functionality wasn't tested to proactively prevent bugs
2. Encourage new pymeasure users to submit PRs if they've tested functionality to remove 'untested' flag

Cons:
1. Adds more code or documentation to codebase
2. Use case doesn't happen often enough to warrant effort.

I think it is worth it, even if it is just standardizing a best practice around documentation. A lot of instruments have addon modules that have documented functionality in the manual but you can't test. Even with an instrument that you have access to, maybe you don't have the necessary testing apparatus to test all functionality.

**Comment by bilderbuchi** [comment]
I opened #747 to track implementation of the `@untested` decorator and related things.
